hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -140,204 +141,205 @@ object InferPType {
-        val elt = coerce[PBaseStruct](coerce[PStreamable](collection.pType2).elementType)
-        PDict(elt.types(0), PArray(elt.types(1)), collection.pType2.required)
+        val elt = coerce[PBaseStruct](coerce[PStreamable](collection._pType2).elementType)
+        PDict(elt.types(0), PArray(elt.types(1)), collection._pType2.required)
       }
       case ArrayMap(a, name, body) => {
         InferPType(a, env)
-        InferPType(body, env.bind(name, a.pType2.asInstanceOf[PArray].elementType))
-        coerce[PStreamable](a.pType2).copyStreamable(body.pType2, body.pType2.required)
+        InferPType(body, env.bind(name, a._pType2.asInstanceOf[PArray].elementType))
+        coerce[PStreamable](a._pType2).copyStreamable(body._pType2, body._pType2.required)
       }
       case ArrayZip(as, names, body, _) =>
-        as.foreach(InferPType(_, env))
-        InferPType(body, env.bindIterable(names.zip(as.map(_.pType2))))
-        coerce[PStreamable](as.head.pType2).copyStreamable(body.pType2, as.forall(_.pType2.required))
+        as.foreach(a => InferPType(a, env))
+
+        InferPType(body, env.bindIterable(names.zip(as.map(_._pType2.asInstanceOf[PArray].elementType))))
+        coerce[PStreamable](as.head._pType2).copyStreamable(body._pType2, as.forall(_._pType2.required))
       case ArrayFilter(a, name, cond) => {
         InferPType(a, env)
-        a.pType2
+        a._pType2
       }
       case ArrayFlatMap(a, name, body) => {
         InferPType(a, env)
-        InferPType(body, env.bind(name, a.pType2.asInstanceOf[PArray].elementType))
+        InferPType(body, env.bind(name, a._pType2.asInstanceOf[PArray].elementType))
 
         // Whether an array must return depends on a, but element requiredeness depends on body (null a elements elided)
-        coerce[PStreamable](a.pType2).copyStreamable(coerce[PIterable](body.pType2).elementType, a.pType2.required)
+        coerce[PStreamable](a._pType2).copyStreamable(coerce[PIterable](body._pType2).elementType, a._pType2.required)
       }
       case ArrayFold(a, zero, accumName, valueName, body) => {
         InferPType(zero, env)
 
         InferPType(a, env)
-        InferPType(body, env.bind(accumName -> zero.pType2, valueName -> a.pType2.asInstanceOf[PArray].elementType))
-        assert(body.pType2 isOfType zero.pType2)
+        InferPType(body, env.bind(accumName -> zero._pType2, valueName -> a._pType2.asInstanceOf[PArray].elementType))
+        assert(body._pType2 isOfType zero._pType2)
 
-        zero.pType2.setRequired(body.pType2.required)
+        zero._pType2.setRequired(body._pType2.required)
       }
       case ArrayFold2(a, acc, valueName, seq, res) =>
         InferPType(a, env)
         acc.foreach { case (_, accIR) => InferPType(accIR, env) }
-        InferPType(a, env)
-        val resEnv = env.bind(acc.map { case (name, accIR) => (name, accIR.pType2)}: _*)
-        val seqEnv = resEnv.bind(valueName -> a.pType2.asInstanceOf[PArray].elementType)
+
```

Comment:
this was a bug; double inference causing our top-level assertion error to trigger (`ir._pType2 == null`)

Reply:
ah, nice.